File: src/main/java/javaclasses/Matrix/Search2dMatrix.java
/*
 * *
 *  * Search a 2d Matrix.java
 *  * Created by Rafsan Ahmad on 1/17/22, 3:21 PM
 *  * Copyright (c) 2022 . All rights reserved.
 *
 */

package javaclasses.Matrix;

public class Search2dMatrix {
    //Leetcode 74
    /*Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following
    properties:

Integers in each row are sorted from left to right.
The first integer of each row is greater than the last integer of the previous row.


Example 1:
    1  | 3  | 5  |7
    10 | 11 | 16 | 20
    23 | 30 | 34 | 60

Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
Output: true

Example 2:

Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
Output: false
*/
    public boolean searchMatrix(int[][] matrix, int target) {
        int len = matrix[0].length;
        for (int i = 0; i < matrix.length; i++) {
            int first = matrix[i][0];
            int last = matrix[i][len - 1];
            if (target >= first && target <= last) {
                //In range
                int left = 0;
                int right = len - 1;
                while (left <= right) {
                    int mid = (left + right) / 2;
                    if (matrix[i][mid] == target) {
                        return true;
                    } else if (matrix[i][mid] <= target) {
                        left = mid + 1;
                    } else {
                        right = mid - 1;
                    }
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Search2dMatrix matrix = new Search2dMatrix();
        int[][] arr = {{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}};
        System.out.println(matrix.searchMatrix(arr, 3));
        System.out.println(matrix.searchMatrix(arr, 13));

        int[][] arr2 = {{1, 3}};
        System.out.println(matrix.searchMatrix(arr2, 3));
    }
}

File: src/main/java/kotlinclasses/BinarySearch/Search2DMatrix.kt
/*
 * *
 *  * Search a 2D Matrix.kt
 *  * Created by Rafsan Ahmad on 6/21/25, 1:33PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.BinarySearch


class Search2DMatrix {
    //https://leetcode.com/problems/search-a-2d-matrix/description/
    /*You are given an m x n integer matrix matrix with the following two properties:

Each row is sorted in non-decreasing order.
The first integer of each row is greater than the last integer of the previous row.
Given an integer target, return true if target is in matrix or false otherwise.

You must write a solution in O(log(m * n)) time complexity.



Example 1:
Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
Output: true

Example 2:
Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
Output: false


Constraints:
m == matrix.length
n == matrix[i].length
1 <= m, n <= 100
-10^4 <= matrix[i][j], target <= 10^4*/

    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        val m = matrix.size
        val n = matrix[0].size

        var row = -1
        for (i in 0 until m) {
            if (matrix[i][0] <= target && matrix[i][n - 1] >= target) {
                row = i
            }
        }

        if (row == -1) return false
        var left = 0
        var right = n

        while (left < right) {
            val mid = (left + right) / 2
            if (matrix[row][mid] == target) {
                return true
            } else if (matrix[row][mid] < target) {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return false
    }
}

fun main() {
    val obj = Search2DMatrix()
    val matrix = arrayOf(
        intArrayOf(1, 3, 5, 7),
        intArrayOf(10, 11, 16, 20),
        intArrayOf(23, 30, 34, 60)
    )

    val matrix2 = arrayOf(
        intArrayOf(1)
    )

    println(obj.searchMatrix(matrix2, 1))
    println(obj.searchMatrix(matrix, 34))
    println(obj.searchMatrix(matrix, 9))
    println(obj.searchMatrix(matrix, 24))
}